pkg/auth/webapp/service2.go

```diff
@@ -564,6 +587,68 @@ func (s *Service2) deriveFinishRedirectURI(session *Session, graph *interaction.
 	return
 }
 
+func (s *Service2) generateSettingsActionGrant(redirectURI string, e *authenticationinfo.Entry) (string, error) {
+	entry, err := s.OAuthSessions.Get(e.OAuthSessionID)
+	if err != nil {
+		return "", err
+	}
+
+	req := entry.T.AuthorizationRequest
+	if req.ResponseType() != "settings_action" {
+		return "", nil
+	}
+
+	result := entry.T.SettingsActionResult
+	if result == nil {
+		return "", nil
+	}
+
+	client := s.OAuthClientResolver.ResolveClient(entry.T.AuthorizationRequest.ClientID())
+
+	// Assume prompt=none for settings action
+	_, uiInfoByProduct, err := s.UIInfoResolver.ResolveForAuthorizationEndpoint(client, req)
+	if err != nil {
+		return "", err
+	}
+	idTokenHintSID := uiInfoByProduct.IDTokenHintSID
+
+	authenticationInfo := e.T
+	autoGrantAuthz := client.IsFirstParty()
+	var authz *oauth.Authorization
+	r := entry.T.AuthorizationRequest
+	if autoGrantAuthz {
+		authz, err = s.Authorizations.CheckAndGrant(
+			r.ClientID(),
+			authenticationInfo.UserID,
+			r.Scope(),
+		)
+	} else {
+		authz, err = s.Authorizations.Check(
+			r.ClientID(),
+			authenticationInfo.UserID,
+			r.Scope(),
+		)
+	}
+	if err != nil {
+		return "", err
+	}
+
+	ss := session.GetSession(s.Request.Context())
+	code, _, err := s.SettingsActionGrantService.CreateSettingsActionGrant(&handler.CreateSettingsActionGrantOptions{
+		Authorization:        authz,
+		IDPSessionID:         ss.SessionID(),
+		AuthenticationInfo:   authenticationInfo,
+		IDTokenHintSID:       idTokenHintSID,
+		RedirectURI:          req.RedirectURI(),
+		AuthorizationRequest: req,
```

Comment:
This is overly complicated. Why do we need to create code here?

Reply:
Should follow ctrl redirect to consent screen during settings action, removed.